### DISCUSSION TITLE
RegCallback example

### DISCUSSION THREAD

**Discussion by monocraft** [Spawnova/ShinsOverlayClass#8] [Q&A] [answered]
Hi Spawnova,
Thank you for sharing the overlay class.  Do you have any example to share with the regcallback?

**Comment by Spawnova** [comment, marked as answer]
Sure, the main purpose of the callbacks is to know when the overlay is being moved, resized or the state changes without having to constantly check

```ahk
#include shinsoverlayclass.ahk

overlay := new ShinsOverlayClass("TITLE_HERE")

overlay.RegCallback(func("_OnSize"),"size") ;callback if size changes
overlay.RegCallback(func("_OnPosition"),"position") ;callback is position changes
overlay.RegCallback(func("_OnActive"),"Active") ;callback if the overlay drawing state changes


loop {
	if (overlay.BeginDraw()) {
		
		overlay.DrawCircle(50,50,10,0xFFFF0000) ;drawing stuff here
		overlay.DrawText("F8 - Exitapp",5,5,24)
		
		overlay.EndDraw()
	}
}

f8::exitapp


_OnActive(overlay,state) {
	tooltip % "Overlay state has changed it is now: " (state ? "Active" : "Inactive")
}
_OnSize(overlay) {
	tooltip % "Overlay size has changed, new size: " overlay.width ", " overlay.height
}
_OnPosition(overlay) {
	tooltip % "Overlay position has changed, new position: " overlay.x ", " overlay.y
}
```

**Comment by monocraft** [comment]
Thank you Spawnova.
I supposed if it call any function it would work if we are registering a mouse click too?
I tried to use your class with lost ark to detect skill that being on cool down.  It seem to detect occasionally only.  Must doing something wrong.
The moving square is what the overlay detecting.
![NVIDIA_Share_iaH9SAcJ3v](https://github.com/Spawnova/ShinsOverlayClass/assets/10327931/f24e7995-2a4c-460b-8f8f-f8e5f55c404e)

**Comment by Spawnova** [comment]
The callbacks are only for interactions with the overlay, it only supports the 3 examples that I listed, You probably need to tighten up how you search for what ability is on cooldown, seems like the image detection may be the issue

**Reply by monocraft** [reply]
You were right.  I didn't realize that there were transparency to the image and it is also animated so it can't find it 100%.  Adjusting with variant seems to help.
![AutoHotkey_GO9HrPhwJY](https://github.com/Spawnova/ShinsOverlayClass/assets/10327931/53af1f68-6a0e-4074-9499-d74f7e1c54e2)

**Comment by monocraft** [comment]
Is there a way to keep the overlay active even if it is in background process?  
Currently if the main client is out of focus, the overlay disappeared.

Still playing with the class.  Just want to say, this overlay class is quite amazing.
![keCAgkg03Q](https://github.com/Spawnova/ShinsOverlayClass/assets/10327931/7d563799-896f-4315-b6cc-5f0ba87f06c7)

**Comment by Spawnova** [comment]
Of course, the 3rd parameter when creating the overlay is a boolean which enabled always showing the overlay even when target window is not in the foreground
```ahk
overlay := new ShinsOverlayClass("TITLE_HERE",1,0) ;title, client space, foreground only disabled
```
alternatively you can set `overlay.attachForeground := 0` to change it any time after instantiating the class

Glad you are having fun with the class =)

**Comment by monocraft** [comment]
Thank you.
